### PR TITLE
Scripts/Spells: Implemented druid eclipse system

### PR DESCRIPTION
**Pull request by matanshukry**
**Changes proposed:**

- Implemented druid eclipse system

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

### CODE REVIEW

File: src/server/scripts/Spells/spell_druid.cpp [hunk 1]
```diff
@@ -1486,6 +1495,136 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+                // Remove stacks from other one as well
+                // reset remaining power on other spellId
+                caster->RemoveAura(cntSpellId);
+                caster->RemoveAura(otherCntSpellId);
+            }
+        }
+    }
+};
+
+// 48517 Eclipse (Solar) + 48518 Eclipse (Luanr)
```
Comment (Faq) [inline src/server/scripts/Spells/spell_druid.cpp:1586]:
```suggestion
// 48517 Eclipse (Solar) + 48518 Eclipse (Lunar)
```

File: src/server/scripts/Spells/spell_druid.cpp [hunk 2]
```diff
@@ -1486,6 +1495,136 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+                    OnSpellCast(SPELL_DRUID_ECLIPSE_SOLAR_SPELL_CNT, SPELL_DRUID_ECLIPSE_LUNAR_SPELL_CNT, SPELL_DRUID_ECLIPSE_SOLAR_AURA);
+                else if (spellInfo->Id == wraith)
+                    OnSpellCast(SPELL_DRUID_ECLIPSE_LUNAR_SPELL_CNT, SPELL_DRUID_ECLIPSE_SOLAR_SPELL_CNT, SPELL_DRUID_ECLIPSE_LUNAR_AURA);
+            }
+        }
+    }
+
+    void HandleApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        SetSpellCount(SPELL_DRUID_ECLIPSE_SOLAR_SPELL_CNT, 2);
```
Comment (Shauren) [inline src/server/scripts/Spells/spell_druid.cpp:1533]: Replace hardcoded 2 with `aurEff->GetAmount()` - there are auras that affect eclipse counting system (334604) and they modify the value with a SPELLMOD_EFFECT1 (already working, you don't need to do anything)

File: src/server/scripts/Spells/spell_druid.cpp [hunk 3]
```diff
@@ -1486,6 +1495,136 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+            {
+                uint32 spellId;
+                if (GetSpellInfo()->Id == SPELL_DRUID_ECLIPSE_SOLAR_AURA)
+                    spellId = SPELL_DRUID_ECLIPSE_LUNAR_SPELL_CNT;
+                else
+                    spellId = SPELL_DRUID_ECLIPSE_SOLAR_SPELL_CNT;
+
+                Aura* aura = player->GetAura(spellId);
+                if (!aura)
+                    player->CastCustomSpell(spellId, SPELLVALUE_AURA_STACK, 2, nullptr, TRIGGERED_FULL_MASK);
```
Comment (Shauren) [inline src/server/scripts/Spells/spell_druid.cpp:1615]: Replace hardcoded 2 by getting AuraEffect* of 79577 and taking its value

File: src/server/scripts/Spells/spell_druid.cpp [hunk 4]
```diff
@@ -1486,6 +1495,136 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+    }
+
+    bool Load() override
+    {
+        return GetCaster()->GetTypeId() == TYPEID_PLAYER;
+    }
+
+    void HandleRemoved(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        if (Player* player = GetCaster()->ToPlayer())
```
Comment (Shauren) [inline src/server/scripts/Spells/spell_druid.cpp:1603]: Don't restrict things to Player only if you aren't using functions from Player class

File: src/server/scripts/Spells/spell_druid.cpp [hunk 5]
```diff
@@ -1486,6 +1495,136 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+        {
+            if (!player->IsInCombat())
```
Comment (Shauren) [inline src/server/scripts/Spells/spell_druid.cpp:1605]: Where are stacks restored in combat after eclipse expires?

Comment (matanshukry) [reply src/server/scripts/Spells/spell_druid.cpp]: Fixed.

File: sql/updates/world/master/2021_02_14_01_world_druid_eclipse.sql [hunk 6]
```diff
@@ -0,0 +1,8 @@
+-- Add eclipse spell script
+DELETE FROM `spell_script_names` WHERE `ScriptName` IN ('spell_dru_eclipse_dummy', 'spell_dru_eclipse_aura', 'spell_dru_eclipse_ooc');
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
+(79577, 'spell_dru_eclipse_dummy'),
+(48517, 'spell_dru_eclipse_aura'),
+(48518, 'spell_dru_eclipse_aura'),
+(329910, 'spell_dru_eclipse_ooc')
+;
```
Comment (mdX7) [inline sql/updates/world/master/2021_02_14_01_world_druid_eclipse.sql:8]: put ; on the last line of the statement, this looks weird :D

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 7]
```diff
@@ -2129,6 +2129,19 @@ void Aura::CallScriptEffectSplitHandlers(AuraEffect* aurEff, AuraApplication con
     }
 }
 
+void Aura::CallScriptOutOfCombatHandlers()
+{
+    for (auto scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
+    {
+        (*scritr)->_PrepareScriptCall(AURA_SCRIPT_HOOK_DISPEL);
```
Comment (mdX7) [inline src/server/game/Spells/Auras/SpellAuras.cpp:2136]: AURA_SCRIPT_HOOK_DISPEL?

File: src/server/scripts/Spells/spell_druid.cpp [hunk 8]
```diff
@@ -1510,6 +1520,198 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+    }
+
+    void Register() override
+    {
+        OnEffectRemove += AuraEffectRemoveFn(spell_dru_eclipse_aura::HandleRemoved, EFFECT_0, SPELL_AURA_ADD_PCT_MODIFIER, AURA_EFFECT_HANDLE_REAL);
+    }
+};
+
+// 329910 - Eclipse out of combat - SPELL_DRUID_ECLIPSE_OCC
+class spell_dru_eclipse_ooc : public AuraScript
```
Comment (mdX7) [inline src/server/scripts/Spells/spell_druid.cpp:1678]: inconsistency - occ vs ooc

File: src/server/scripts/Spells/spell_druid.cpp [hunk 9]
```diff
@@ -1510,6 +1520,198 @@ class spell_dru_wild_growth : public SpellScriptLoader
...
+    PrepareAuraScript(spell_dru_eclipse_aura);
+
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({ SPELL_DRUID_ECLIPSE_LUNAR_SPELL_CNT, SPELL_DRUID_ECLIPSE_SOLAR_SPELL_CNT, SPELL_DRUID_ECLIPSE_DUMMY });
+    }
+
+    bool Load() override
+    {
+        return GetCaster()->GetTypeId() == TYPEID_PLAYER;
```
Comment (mdX7) [inline src/server/scripts/Spells/spell_druid.cpp:1641]: owner / nullcheck

always keep in mind: caster may be null in AuraScripts. Imagine i'm dotting you with moonfire and logout after it => caster null.

**Comment by illfated** [comment]
Thank you for the clean-up. Looks quite readable now.

**Comment by Aokromes** [comment]
Conflicting files
src/server/scripts/Spells/spell_druid.cpp

**Comment by illfated** [comment]
Likely due to the recent merge of PR #26111 .

**Review by ghost — APPROVED** [review]

**Comment by Shauren** [comment]
Fixups done: 
* Applying counters on login (wasnt working since implementation of new aura interrupt flags)
* Created spell_proc entry to override SpellPhaseMask, counter should decrement when spell missile launches, not hits target
* Changed used script hooks - casting spells and removing auras is not allowed in On* hooks, that must be done in After*
* Additional counters out of combat reset rule, only refilling when any counter is completely empty (verified on retail)
* UnitOwner -> Target

**Pull request merged**